File: services/flow-engine/sync-bridge.ts
/**
 * SyncBridge — Controle da ponte síncrona com harvest
 *
 * Modelo Harvest + Barrier:
 *   - Coletar todos os nós leves (text, reaction, interactive) até barreira
 *   - Combinar em um único payload sync (button_reaction com mapped)
 *   - Após barreira (MEDIA/DELAY), tudo é async via API
 *
 * @see docs/interative_message_flow_builder.md §14.2
 */

import log from '@/lib/log';
import type { SynchronousResponse } from '@/types/flow-engine';

/** Componentes coletados para o payload sync combinado */
interface HarvestedComponents {
  emoji?: string;
  targetMessageId?: string;
  /** ID da mensagem de contexto (para reply em contexto, mesmo sem emoji) */
  contextMessageId?: string;
  texts: string[];
  interactive?: Record<string, unknown>;
}

export class SyncBridge {
  private syncPayload: SynchronousResponse | null = null;
  private syncConsumed: boolean = false;

  /** Reação pendente para combinar com o próximo texto (formato button_reaction) */
  private pendingReaction: { emoji: string; targetMessageId: string } | null = null;

  /** Componentes coletados durante harvest */
  private harvestedComponents: HarvestedComponents = { texts: [] };

  // ---------------------------------------------------------------------------
  // Sync Payload (apenas 1, a primeira mensagem interativa)
  // ---------------------------------------------------------------------------

  /**
   * Ainda podemos usar a ponte síncrona?
   * TRUE apenas se ainda não enviamos nada sync.
   */
  canSync(): boolean {
    return !this.syncConsumed && this.syncPayload === null;
  }

  /**
   * Define o payload para resposta síncrona.
   * Se já tiver payload, ignora (só a primeira conta).
   */
  setSyncPayload(payload: SynchronousResponse): void {
    if (this.syncConsumed) {
      log.debug('[SyncBridge] setSyncPayload ignorado — ponte já foi usada');
      return;
    }
    if (this.syncPayload !== null) {
      log.debug('[SyncBridge] setSyncPayload ignorado — já tem payload pendente');
      return;
    }
    this.syncPayload = payload;
    log.debug('[SyncBridge] Payload sync definido');
  }

  /**
   * Tem payload pendente para enviar na ponte?
   */
  hasSyncPayload(): boolean {
    return this.syncPayload !== null;
  }

  /**
   * Consome o payload sync (para enviar na resposta HTTP).
   * Marca a ponte como usada — depois disso, tudo é async.
   */
  consumeSyncPayload(): SynchronousResponse | null {
    if (this.syncPayload === null) {
      return null;
    }
    const payload = this.syncPayload;
    this.syncPayload = null;
    this.syncConsumed = true;
    log.debug('[SyncBridge] Payload sync consumido — ponte fechada');
    return payload;
  }

  /**
   * A ponte já foi usada?
   */
  isBridgeClosed(): boolean {
    return this.syncConsumed;
  }

  // ---------------------------------------------------------------------------
  // Pending Reaction (para combinar REACTION + TEXT em um único payload)
  // ---------------------------------------------------------------------------

  /**
   * Armazena uma reação pendente para combinar com o próximo texto.
   */
  setPendingReaction(emoji: string, targetMessageId: string): void {
    this.pendingReaction = { emoji, targetMessageId };
    log.debug('[SyncBridge] Reação pendente armazenada', { emoji, targetMessageId });
  }

  /**
   * Consome e retorna a reação pendente (se houver).
   */
  consumePendingReaction(): { emoji: string; targetMessageId: string } | null {
    const reaction = this.pendingReaction;
    this.pendingReaction = null;
    return reaction;
  }

  /**
   * Tem reação pendente?
   */
  hasPendingReaction(): boolean {
    return this.pendingReaction !== null;
  }

  // ---------------------------------------------------------------------------
  // Harvest Components (para combinar múltiplos itens no sync)
  // ---------------------------------------------------------------------------

  /**
   * Define o message_id de contexto (wamid do botão clicado).
   * Usado como fallback quando não há REACTION para fornecer targetMessageId.
   */
  setContextMessageId(messageId: string): void {
    this.harvestedComponents.contextMessageId = messageId;
    log.debug('[SyncBridge] Context message_id definido', { messageId });
  }

  /**
   * Adiciona texto ao harvest.
   */
  addHarvestedText(text: string): void {
    this.harvestedComponents.texts.push(text);
    log.debug('[SyncBridge] Texto adicionado ao harvest', {
      textPreview: text.slice(0, 50),
      totalTexts: this.harvestedComponents.texts.length,
    });
  }

  /**
   * Define o emoji do harvest.
   */
  setHarvestedEmoji(emoji: string, targetMessageId: string): void {
    this.harvestedComponents.emoji = emoji;
    this.harvestedComponents.targetMessageId = targetMessageId;
    log.debug('[SyncBridge] Emoji adicionado ao harvest', { emoji });
  }

  /**
   * Define a mensagem interativa do harvest.
   */
  setHarvestedInteractive(payload: Record<string, unknown>): void {
    this.harvestedComponents.interactive = payload;
    log.debug('[SyncBridge] Interactive adicionado ao harvest');
  }

  /**
   * Tem conteúdo harvested pendente?
   */
  hasHarvestedContent(): boolean {
    return (
      this.harvestedComponents.texts.length > 0 ||
      !!this.harvestedComponents.emoji ||
      !!this.harvestedComponents.interactive
    );
  }

  /**
   * Constrói e retorna o payload sync combinado (formato legado button_reaction).
   * Combina: emoji + textos + interactive em um único JSON.
   *
   * Formato de saída:
   * {
   *   action_type: 'button_reaction',
   *   emoji: '❤️',
   *   text: 'texto combinado',
   *   whatsapp: { message_id, reaction_emoji, response_text },
   *   mapped: { whatsapp: { type: 'interactive', interactive: {...} } }
   * }
   */
  buildCombinedPayload(channel: string): SynchronousResponse | null {
    if (!this.hasHarvestedContent() && !this.pendingReaction) {
      return null;
    }

    // Usar pending reaction se não tiver harvested emoji
    const emoji = this.harvestedComponents.emoji ?? this.pendingReaction?.emoji;
    const targetMessageId = this.harvestedComponents.targetMessageId
      ?? this.pendingReaction?.targetMessageId
      ?? this.harvestedComponents.contextMessageId;

    // Combinar todos os textos
    const combinedText = this.harvestedComponents.texts.join('\n\n');

    // Sem nada para enviar?
    if (!emoji && !combinedText && !this.harvestedComponents.interactive) {
      return null;
    }

    // Construir payload base
    const payload: Record<string, unknown> = {
      action_type: 'button_reaction',
    };

    if (emoji) payload.emoji = emoji;
    if (combinedText) payload.text = combinedText;

    // Construir channel payload (whatsapp/instagram/facebook)
    const channelPayload: Record<string, unknown> = {};
    if (targetMessageId) channelPayload.message_id = targetMessageId;
    if (emoji) channelPayload.reaction_emoji = emoji;
    if (combinedText) channelPayload.response_text = combinedText;

    if (Object.keys(channelPayload).length > 0) {
      const channelKey = channel === 'instagram' ? 'instagram'
        : channel === 'facebook' ? 'facebook'
        : 'whatsapp';
      payload[channelKey] = channelPayload;
    }

    // Adicionar interactive se existir (no formato mapped)
    if (this.harvestedComponents.interactive) {
      const channelKey = channel === 'instagram' ? 'instagram'
        : channel === 'facebook' ? 'facebook'
        : 'whatsapp';

      payload.mapped = {
        [channelKey]: {
          type: 'interactive',
          interactive: this.harvestedComponents.interactive,
        },
      };
    }

    log.debug('[SyncBridge] Payload combinado construído', {
      hasEmoji: !!emoji,
      hasText: !!combinedText,
      hasInteractive: !!this.harvestedComponents.interactive,
    });

    // Limpar harvest após construir
    this.clearHarvest();

    return payload as SynchronousResponse;
  }

  /**
   * Limpa os componentes harvested.
   */
  clearHarvest(): void {
    this.harvestedComponents = { texts: [] };
    this.pendingReaction = null;
  }
}

File: platform-backend/domains/socialwise/services/flow/sync_bridge.py
"""Synchronous bridge for the first flow response returned to Chatwit."""

from __future__ import annotations

from typing import Any


class SyncBridge:
    def __init__(self, force_async: bool = False) -> None:
        self._sync_payload: dict[str, Any] | None = None
        self._sync_consumed = force_async
        self._pending_reaction: dict[str, str] | None = None
        self._context_message_id: str | None = None
        self._harvest_texts: list[str] = []
        self._harvest_emoji: str | None = None
        self._harvest_target_message_id: str | None = None
        self._harvest_interactive: dict[str, Any] | None = None

    def can_sync(self) -> bool:
        return not self._sync_consumed and self._sync_payload is None

    def set_sync_payload(self, payload: dict[str, Any]) -> None:
        if self._sync_consumed or self._sync_payload is not None:
            return
        self._sync_payload = payload

    def has_sync_payload(self) -> bool:
        return self._sync_payload is not None

    def consume_sync_payload(self) -> dict[str, Any] | None:
        payload = self._sync_payload
        self._sync_payload = None
        if payload is not None:
            self._sync_consumed = True
        return payload

    def is_bridge_closed(self) -> bool:
        return self._sync_consumed

    def close_sync_window(self) -> None:
        self._sync_consumed = True

    def set_pending_reaction(self, emoji: str, target_message_id: str) -> None:
        self._pending_reaction = {"emoji": emoji, "target_message_id": target_message_id}

    def consume_pending_reaction(self) -> dict[str, str] | None:
        reaction = self._pending_reaction
        self._pending_reaction = None
        return reaction

    def has_pending_reaction(self) -> bool:
        return self._pending_reaction is not None

    def set_context_message_id(self, message_id: str) -> None:
        self._context_message_id = message_id

    def add_harvested_text(self, text: str) -> None:
        if text:
            self._harvest_texts.append(text)

    def set_harvested_emoji(self, emoji: str, target_message_id: str) -> None:
        self._harvest_emoji = emoji
        self._harvest_target_message_id = target_message_id

    def set_harvested_interactive(self, payload: dict[str, Any]) -> None:
        self._harvest_interactive = payload

    def has_harvested_content(self) -> bool:
        return bool(self._harvest_texts or self._harvest_emoji or self._harvest_interactive)

    def build_combined_payload(self, channel: str) -> dict[str, Any] | None:
        if not self.has_harvested_content() and not self._pending_reaction:
            return None

        emoji = self._harvest_emoji or (self._pending_reaction or {}).get("emoji")
        target_message_id = (
            self._harvest_target_message_id
            or (self._pending_reaction or {}).get("target_message_id")
            or self._context_message_id
        )
        combined_text = "\n\n".join(part for part in self._harvest_texts if part)
        if not emoji and not combined_text and not self._harvest_interactive:
            return None

        payload: dict[str, Any] = {"action_type": "button_reaction"}
        if emoji:
            payload["emoji"] = emoji
        if combined_text:
            payload["text"] = combined_text

        channel_key = "facebook" if channel == "facebook" else "instagram" if channel == "instagram" else "whatsapp"
        channel_payload: dict[str, Any] = {}
        if target_message_id:
            channel_payload["message_id"] = target_message_id
        if emoji:
            channel_payload["reaction_emoji"] = emoji
        if combined_text:
            channel_payload["response_text"] = combined_text
        if channel_payload:
            payload[channel_key] = channel_payload

        if self._harvest_interactive:
            payload["mapped"] = {
                channel_key: {
                    "type": "interactive",
                    "interactive": self._harvest_interactive,
                }
            }
        return payload
